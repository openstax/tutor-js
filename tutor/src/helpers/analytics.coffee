_ = require 'underscore'

Routes = require '../routes'

{CurrentUserStore} = require '../flux/current-user'

DestinationHelper  = require '../helpers/routes-and-destinations'

# generate custom event data for routes
Events =
  viewTaskStep: ({courseId}) ->
    # track all work done on a course
    Analytics.sendEvent 'Course', 'Work', label: courseId

  viewStudentDashboard: ({courseId}) ->
    # compare activity between courses
    Analytics.sendEvent 'Student', 'Dashboard', label: courseId


# Translators convert a url like '/foo/bar/123/baz/1' into a simplified one like just '/foo/bar'
Translators =

  dashboard:            ({courseId}) -> "/student/choose-course/#{courseId}"
  viewPractice:         ({courseId}) -> "/student/practice/#{courseId}"
  viewPerformanceForecast:  ({courseId}) -> "/student/performance-forecast/#{courseId}"
  viewStudentDashboard: ({courseId}) -> "/student/dashboard/#{courseId}"

  calendarByDate:        ({courseId}) -> "/teacher/calendar/#{courseId}"
  viewTeacherPerformanceForecast:      ({courseId}) -> "/teacher/performance-forecast/#{courseId}"
  viewScores:            ({courseId}) -> "/teacher/student-scores/#{courseId}"
  courseSettings:        ({courseId}) -> "/teacher/roster/#{courseId}"
  editReading:           ({courseId}) -> "/teacher/assignment/edit/reading/#{courseId}"
  editHomework:          ({courseId}) -> "/teacher/assignment/edit/homework/#{courseId}"
  editExternal:          ({courseId}) -> "/teacher/assignment/edit/external/#{courseId}"
  createReading:         ({courseId}) -> "/teacher/assignment/create/reading/#{courseId}"
  createHomework:        ({courseId}) -> "/teacher/assignment/create/homework/#{courseId}"
  createExternal:        ({courseId}) -> "/teacher/assignment/create/external/#{courseId}"
  calendarViewPlanStats: ({courseId}) -> "/teacher/metrics/quick/#{courseId}"
  reviewTask:            ({courseId}) -> "/teacher/metrics/review/#{courseId}"

  viewReferenceBook:        ({courseId}) -> "/reference-view/#{courseId}"
  viewReferenceBookSection: ({courseId}) -> "/reference-view/#{courseId}"
  viewReferenceBookPage:    ({courseId}) -> "/reference-view/#{courseId}"

  # Task steps are viewed by both teacher and student with no difference in params
  viewTaskStep:         ({courseId}) ->
    role = CurrentUserStore.getCourseRole(courseId, true)
    "/#{role}/task-step/#{courseId}"


GA = undefined

Analytics =

  setTracker: (tracker) -> GA = tracker

  sendPageView: (url) ->
    GA?('send', 'pageview', url)

  onNavigation: (path) ->
    return unless GA

    route = Routes.pathToEntry(path)

    return @sendPageView("/not-found/#{change.path}") unless route

    path = Translators[route.entry.name]?( route.match.params ) or route.match.pathname

    # if we're also going to send custom events then we set the page
    if Events[route.entry.name]
      GA('set', 'page', path)
      Events[route.entry.name]( route.match.params )
      @sendPageView() # url's not needed since it was set before events
    else
      @sendPageView(route.match.pathname)

  sendEvent: (category, action, attrs) ->
    return unless GA
    GA('send', 'event', category, action, attrs.label, attrs.value)

module.exports = Analytics
