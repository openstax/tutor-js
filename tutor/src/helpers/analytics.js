import { partial, get } from 'lodash';

import Router from '../helpers/router';
import Courses from '../models/courses-map';

// generate custom event data for routes
const Events = {
  viewTaskStep({ courseId }) {
    // track all work done on a course
    return Analytics.sendEvent('Course', 'Work', { label: courseId });
  },

  viewStudentDashboard({ courseId }) {
    // compare activity between courses
    return Analytics.sendEvent('Student', 'Dashboard', { label: courseId });
  },
};

// a bit shorter helper methods
const isTeacher = courseId => get(Courses.get(courseId), 'currentRole.isTeacher', false);
const getRole = function(courseId) {
  if (Courses.get(courseId).is_teacher) { return 'teacher'; } else { return 'student'; }
};

const assignmentTypeTranslator = function(assignmentType, { courseId, id }) {
  const newOrEdit = id === 'new' ? 'create' : 'edit';
  return `/teacher/assignment/${newOrEdit}/${assignmentType}/${courseId}`;
};

function viewReferenceBook({ courseId, pageId }) {
  const url = `/reference-view/${courseId}`;
  if (!pageId) {
    return url;
  }
  const course = Courses.get(courseId);
  const page = course.referenceBook.pages.byId.get(pageId);
  return page ? `${url}/section/${page.chapter_section.key}` : url;
}

// Translators convert a url like '/foo/bar/123/baz/1' into a simplified one like just '/foo/bar'
const Translators = {
  dashboard({ courseId }) {
    if (isTeacher(courseId)) { return `/teacher/calendar/${courseId}`; } else { return `/student/dashboard/${courseId}`; }
  },
  practiceTopics({ courseId }) { return `/student/practice/${courseId}`; },
  viewPerformanceGuide({ courseId }) {
    return `/${getRole(courseId)}/performance-forecast/${courseId}`;
  },
  calendarByDate({ courseId }) { return `/teacher/calendar/${courseId}`; },
  viewGradebook({ courseId }) { return `/teacher/student-scores/${courseId}`; },
  courseSettings({ courseId }) { return `/teacher/roster/${courseId}`; },
  editAssignment: partial(assignmentTypeTranslator, 'assignment'),
  calendarViewPlanStats({ courseId }) { return `/teacher/metrics/quick/${courseId}`; },
  reviewTask({ courseId }) { return `/teacher/metrics/review/${courseId}`; },
  viewReferenceBookSection: viewReferenceBook,
  viewReferenceBookPage: viewReferenceBook,
  viewReferenceBook,

  // Task steps are viewed by both teacher and student with no difference in params
  viewTaskStep({ courseId }) {
    const course = Courses.get(courseId);
    const role = course ? course.primaryRole.type : 'unknown';
    return `/${role}/task-step/${courseId}`;
  },
};

const TAG_MANAGER_ID = 'GTM-W6N7PB';

let GA = undefined;
let trackerNames = undefined;

var Analytics = {
  // copied from BIT
  initTagManager(w, d, s, l, i) {
    // Disable ESLint rules since we're copying Google's script
    /* eslint-disable one-var, prefer-const, prefer-template */
    w[l] = w[l] || [];
    w[l].push({
      'gtm.start': new Date().getTime(),
      event: 'gtm.js',
    });
    let f = d.getElementsByTagName(s)[0], j = d.createElement(s), dl = l !== 'dataLayer' ? '&l=' + l : '';
    // Breaks in tests because there are no scripts
    if (f) {
      j.async = true;
      j.src = '//www.googletagmanager.com/gtm.js?id=' + i + dl;
      f.parentNode.insertBefore(j, f);
    }
  },
  
  // Set the Command Queue (ga) and cache the (possibly multiple) tracker names
  setGa(ga) {
    trackerNames = undefined;
    GA = ga;
    if (GA) {
      GA(function() {
        trackerNames = GA.getAll().map(function(tracker) { return tracker.get('name'); });
      });
      this.initTagManager(window, document, 'script', 'dataLayer', TAG_MANAGER_ID);
    }

    return GA;
  },

  // Calls all trackers with the command and arguments given here
  ga(command, ...params) {
    if (!GA || !trackerNames) { return; }
    trackerNames.forEach((trackerName) => {
      this.realGa(trackerName + '.' + command, ...params);
    });
  },

  // Exists largely for spec mocking
  realGa(...args) {
    GA(...args);
  },

  sendPageView(url) {
    this.ga('send', 'pageview', url);
  },

  recordCourseDimension(courseId) {
    const course = Courses.get(courseId);
    if (course) {
      this.ga('set', 'dimension1', course.appearance_code);
    }
  },

  onNavigation(path) {
    if (!GA) { return; }
    const route = Router.currentMatch(path);

    if (!route) {
      this.sendPageView(`/not-found${path}`);
      return;
    }
    const { courseId } = route.params;
    if (courseId) { this.recordCourseDimension(courseId); }
    const translatedPath = Translators[route.entry.name] ?
      Translators[route.entry.name]( route.params ) : route.pathname;

    this.ga('set', 'page', translatedPath);
    // if we're also going to send custom events then we set the page
    if (Events[route.entry.name]) {
      Events[route.entry.name]( route.params );
    }

    this.sendPageView(); // url's not needed since it was set before events
  },

  sendEvent(category, action, attrs = {}) {
    this.ga('send', 'event', category, action, attrs.label, attrs.value);
  },
};

export default Analytics;
