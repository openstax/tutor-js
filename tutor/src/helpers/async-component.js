import React from 'react';
import AsyncErrorBoundary from '../components/error-monitoring/async-load-error';
import LoadingAnimation from '../../../shared/src/components/loading-animation'

export function asyncComponent(loader, name = '') {

    const Component = React.lazy(loader);
    const loading = <LoadingAnimation message={`Loading ${name}â€¦`} />;

    const Loader = (props) => (
        <AsyncErrorBoundary>
            <React.Suspense fallback={loading}>
                <Component {...props} />
            </React.Suspense>
        </AsyncErrorBoundary>
    );

    return (props) => <Loader {...props} />;
}


export function loadAsync(resolve, name) {
    // return a function so the router will only evaluate it when it's needed
    // in the future we may insert role dependant logic here
    return () => asyncComponent(resolve, name);
}
