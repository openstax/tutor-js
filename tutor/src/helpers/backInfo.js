import { navHistory } from '../models'


const backInfo = () => {
    // Gets route to last path that was not the same as the current one
    // See TransitionStore for more detail.
    const historyInfo = navHistory.latest

    const text = historyInfo != null && !!historyInfo.name ? `Back to ${historyInfo.name}` : 'My Courses';
    const to =  historyInfo != null && !!historyInfo.path ? historyInfo.path : 'myCourses';

    return {
        text,
        to,
    };
};


export default backInfo;
