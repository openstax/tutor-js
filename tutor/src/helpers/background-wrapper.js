import { styled } from 'vendor';
import { navbars } from '../theme.js';

const BackgroundWrapper = styled.div`
  background: #fff;
  min-height: calc(100vh - ${navbars.top.height} - ${navbars.bottom.height});
  position: relative;
  overflow: hidden;
  padding: 0 2.4rem;
`;

const ContentWrapper = styled.div`
  max-width: 1200px;
  margin: 0 auto;
`;

export { BackgroundWrapper, ContentWrapper };
