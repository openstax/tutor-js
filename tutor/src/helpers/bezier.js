function EASE_IN_OUT(t) {
    if (t < .5) { return 4 * t * t * t; } else { return ((t - 1) * ((2 * t) - 2) * ((2 * t) - 2)) + 1; }
}

function CALC(start, end, elapsed, duration) {
    if (elapsed > duration) { return end; }
    return start + ((end - start) * EASE_IN_OUT(elapsed / duration));
}

export default function bezier({
    range, duration, onStep, onComplete,
}) {
    const startTime = Date.now();
    const requestAnimationFrame = window.requestAnimationFrame || window.setTimeout;

    const step = () => {
        const elapsed = Date.now() - startTime;
        onStep(CALC(range[0], range[1], elapsed, duration));
        if (elapsed < duration) {
            requestAnimationFrame(step);
        } else if (onComplete) {
            onComplete();
        }
    };

    step();
}
