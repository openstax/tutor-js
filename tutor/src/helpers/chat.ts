import { Raven } from '../models'

const DEPLOYMENT_ID   = '5724P000000ghju';
const ORGANIZATION_ID = '00DU0000000Kwch';
const ELEMENT_ID      = '5736f000000k9eQ';


const win = window as any

export const Chat = {
    initialized: false,

    get isEnabled() {
        return Boolean(win.liveagent);
    },

    get isOnline() {
        if (!this.isEnabled) { return false; }
        const chat = document.querySelector<HTMLDivElement>('.chat.enabled');
        return chat && chat.style.display !== 'none';
    },

    initialize() {
    // will emit debugging info to the console
    // liveagent.enableLogging();
    },

    setElementVisiblity(online: any, offline?: any) {
        if (!this.isEnabled) { return; }
        try {
            win.liveagent.showWhenOnline(ELEMENT_ID, online);
            if (offline) {
                win.liveagent.showWhenOffline(ELEMENT_ID, offline);
            }
            if (!this.initialized) {
                win.liveagent.init('https://d.la1-c2-ia5.salesforceliveagent.com/chat', DEPLOYMENT_ID, ORGANIZATION_ID);
                this.initialized = true;
            }
        } catch(err) {
            Raven.captureException(err);
        }
    },

    start() {
        win.liveagent.startChat(ELEMENT_ID);
    },

};
