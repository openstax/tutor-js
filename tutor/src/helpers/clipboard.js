import _ from 'underscore';

const SUPPORTED = typeof document.queryCommandSupported === 'function' ? document.queryCommandSupported('copy') : undefined;

export default {
    isSupported: _.constant(!!SUPPORTED),

    copy() {
        if (!SUPPORTED) { return; }

        try {
            document.execCommand('copy');
        } catch (e) {
            // eslint-disable-next-line no-console
            console.warn('clipboard copy failed', e);
        }
    },
};
