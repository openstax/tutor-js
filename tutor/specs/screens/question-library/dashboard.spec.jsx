import { C } from '../../helpers';
import Factory, { FactoryBot } from '../../factories';
import { last } from 'lodash';
import Dashboard from '../../../src/screens/question-library/dashboard';
import ExerciseHelpers from '../../../src/helpers/exercise';
import ScrollTo from '../../../src/helpers/scroll-to';

jest.mock('../../../../shared/src/components/html', () => ({ html }) =>
  html ? <div dangerouslySetInnerHTML={{ __html: html }} /> : null
);
jest.mock('../../../src/helpers/exercise');
jest.mock('../../../src/helpers/scroll-to');

describe('Questions Dashboard Component', function() {
  let props, course, exercises, book, page_ids;

  beforeEach(function() {
    course = Factory.course();
    book = course.referenceBook;
    course.referenceBook.onApiRequestComplete({ data: [FactoryBot.create('Book')] });
    exercises = Factory.exercisesMap({ ecosystem_id: course.ecosystem_id, pageIds: [], count: 0 });

    exercises.fetch = jest.fn(() => Promise.resolve());
    page_ids = course.referenceBook.children[1].children.assignable.map(p => p.id);
    const items = page_ids.map(page_id =>
      FactoryBot.create(
        'TutorExercise',
        {
          // QL displays homework assignment first
          pool_types: ['homework_core'],
          page_uuid: book.pages.byId.get(page_id).uuid,
        },
      ),
    );
    exercises.onLoaded({ data: { items } }, [{ book, page_ids }]);
    props = {
      course,
      exercises,
    };
  });

  const displayExercises = () => {
    const dash = mount(<C><Dashboard {...props} /></C>);
    dash.find('.chapter-checkbox button').at(1).simulate('click');
    dash.find('.section-controls .btn-primary').simulate('click');
    return dash;
  };

  it('fetches and displays', () => {
    const dash = mount(<C><Dashboard {...props} /></C>);
    expect(dash).not.toHaveRendered('NoExercisesFound');
    dash.find(`div[data-section-id="${page_ids[0]}"]`).simulate('click');
    dash.find('.section-controls .btn-primary').simulate('click');
    expect(exercises.fetch).toHaveBeenCalledWith({
      course: course, limit: false, page_ids: [page_ids[0]],
    });
    expect(dash).not.toHaveRendered('NoExercisesFound');
    dash.unmount();
  });

  it('renders exercise details', () => {
    const dash = displayExercises();
    expect(dash).not.toHaveRendered('NoExercisesFound');
    const uid = dash.find('[data-exercise-id]').last().prop('data-exercise-id');
    dash.find(`[data-exercise-id="${uid}"] .action.details`).simulate('click');
    expect(dash).toHaveRendered('.exercise-details');
    dash.find('a.show-cards').simulate('click');
    expect(dash).toHaveRendered('.exercise-cards');
    expect(last(ScrollTo.mock.instances).scrollToSelector).toHaveBeenCalledWith(
      `[data-exercise-id="${uid}"]`, { immediate: true }
    );
    dash.unmount();
  });

  it('renders details, then select more sections, then display', () => {
    const dash = mount(<C><Dashboard {...props} /></C>);
    expect(dash).toHaveRendered('SectionsChooser');
    const input = `[data-section-id="${page_ids[1]}"] Icon`;
    // square is not selected
    expect(dash.find(input).first().props().type).toBe('square');
    dash.find(input).first().simulate('click');
    dash.find('.section-controls .btn-primary').simulate('click');
    expect(dash).toHaveRendered('ExercisesDisplay');
    dash.find('.action.details').at(0).simulate('click');
    expect(dash).toHaveRendered('ExerciseDetails');
    dash.find('ExercisesDisplay').props().onSelectSections();
    expect(dash).toHaveRendered('SectionsChooser');
    // check-square-solid is selected
    expect(dash.find(input).first().props().type).toBe('check-square-solid');
    dash.find('.section-controls .btn-primary').simulate('click');
    expect(dash).toHaveRendered('ExercisesDisplay');
    expect(dash).not.toHaveRendered('ExerciseDetails');
    dash.unmount();
  });

  it('can exclude exercises', () => {
    const dash = displayExercises();
    expect(dash).not.toHaveRendered('NoExercisesFound');
    dash.find('.action.details').at(0).simulate('click');
    expect(dash).toHaveRendered('.exercise-details');
    course.saveExerciseExclusion = jest.fn();
    const uid = dash.find('[data-exercise-id]').last().prop('data-exercise-id');
    const exercise = exercises.array.find(e => uid == e.content.uid);
    dash.find(`[data-exercise-id="${uid}"] .action.exclude`).simulate('click');
    expect(course.saveExerciseExclusion).toHaveBeenCalledWith({ exercise, is_excluded: true });
    dash.unmount();
  });

  it('can report errors', () => {
    const dash = displayExercises();
    expect(dash).not.toHaveRendered('NoExercisesFound');
    dash.find('.action.details').at(0).simulate('click');
    dash.find('.action.report-error').simulate('click');
    const uid = dash.find('[data-exercise-id]').last().prop('data-exercise-id');
    const exercise = exercises.array.find(e => uid == e.content.uid);
    expect(ExerciseHelpers.openReportErrorPage).toHaveBeenCalledWith(exercise, course);
    dash.unmount();
  });

  it('clears when cancel is clicked', () => {
    const dash = mount(<C><Dashboard {...props} /></C>);
    dash.find('.chapter-checkbox button').at(1).simulate('click');
    expect(dash.find('.section.selected').length).toBeGreaterThan(10);
    dash.find('button.cancel').simulate('click');
    expect(dash.find('.section.selected')).toHaveLength(0);
    dash.unmount();
  });

});
