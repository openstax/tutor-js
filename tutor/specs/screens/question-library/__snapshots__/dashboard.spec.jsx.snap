// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions Dashboard Component matches snapshot 1`] = `
<div
  className="questions-dashboard"
>
  <div
    className="sections-chooser panel"
  >
    <div
      className="header"
    >
      <div
        className="wrapper"
      >
        <h2>
          Question Library
        </h2>
        <a
          className="btn btn-default"
          href="/course/1"
          onClick={[Function]}
        >
          Back to Dashboard
        </a>
      </div>
    </div>
    <div
      className="sections-list"
      data-tour-region-id="question-library-sections-chooser"
    >
      <div
        className="sections-chooser"
      >
        <div
          className="panel-group"
          role="tablist"
        >
          <div
            className="panel panel-default"
            id={null}
          >
            <div
              className="panel-heading"
            >
              <div
                className="chapter-heading panel-title"
                data-chapter-section={1}
              >
                <a
                  aria-controls={undefined}
                  aria-expanded={true}
                  aria-selected={true}
                  className={null}
                  href={undefined}
                  onClick={[Function]}
                  role="tab"
                >
                  <span
                    className="chapter-checkbox"
                  >
                    <span
                      className="tri-state-checkbox unchecked"
                      onClick={[Function]}
                      tabIndex={1}
                    >
                      <i
                        className="tutor-icon fa fa-square-o clickable"
                        onClick={[Function]}
                        role="button"
                        style={
                          Object {
                            "cursor": "pointer",
                          }
                        }
                        type="square-o"
                      />
                    </span>
                  </span>
                  <span
                    className="chapter-number"
                  >
                    Chapter 
                    <span
                      className="chapter-section"
                      data-chapter-section="1"
                    >
                      1
                    </span>
                     - 
                  </span>
                  <span
                    className="chapter-title"
                  >
                     
                    Kinematics
                     
                  </span>
                  <div
                    className="browse-the-book"
                    onClick={[Function]}
                    role="link"
                  >
                    Browse this Chapter
                  </div>
                </a>
              </div>
            </div>
            <div
              aria-expanded={null}
              aria-hidden={false}
              className="panel-collapse collapse in"
              id={undefined}
              role="tabpanel"
            >
              <div
                className="panel-body"
              >
                <div
                  className="section"
                  data-page-id="234"
                  onClick={[Function]}
                >
                  <span
                    className="section-checkbox"
                  >
                    <input
                      checked={false}
                      readOnly={true}
                      type="checkbox"
                    />
                  </span>
                  <span
                    className="chapter-section"
                    data-chapter-section="1.1"
                  >
                    1.1
                  </span>
                  <span
                    className="section-title"
                  >
                     
                    Kinematics in 1 dim
                  </span>
                </div>
                <div
                  className="section"
                  data-page-id="235"
                  onClick={[Function]}
                >
                  <span
                    className="section-checkbox"
                  >
                    <input
                      checked={false}
                      readOnly={true}
                      type="checkbox"
                    />
                  </span>
                  <span
                    className="chapter-section"
                    data-chapter-section="1.2"
                  >
                    1.2
                  </span>
                  <span
                    className="section-title"
                  >
                     
                    Kinematics in 2 dim
                  </span>
                </div>
                <div
                  className="section"
                  data-page-id="236"
                  onClick={[Function]}
                >
                  <span
                    className="section-checkbox"
                  >
                    <input
                      checked={false}
                      readOnly={true}
                      type="checkbox"
                    />
                  </span>
                  <span
                    className="chapter-section"
                    data-chapter-section="1.3"
                  >
                    1.3
                  </span>
                  <span
                    className="section-title"
                  >
                     
                    Kinematics in 3 dim
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="panel-group"
          role="tablist"
        >
          <div
            className="panel panel-default"
            id={null}
          >
            <div
              className="panel-heading"
            >
              <div
                className="chapter-heading panel-title"
                data-chapter-section={2}
              >
                <a
                  aria-controls={undefined}
                  aria-expanded={false}
                  aria-selected={false}
                  className="collapsed"
                  href={undefined}
                  onClick={[Function]}
                  role="tab"
                >
                  <span
                    className="chapter-checkbox"
                  >
                    <span
                      className="tri-state-checkbox unchecked"
                      onClick={[Function]}
                      tabIndex={1}
                    >
                      <i
                        className="tutor-icon fa fa-square-o clickable"
                        onClick={[Function]}
                        role="button"
                        style={
                          Object {
                            "cursor": "pointer",
                          }
                        }
                        type="square-o"
                      />
                    </span>
                  </span>
                  <span
                    className="chapter-number"
                  >
                    Chapter 
                    <span
                      className="chapter-section"
                      data-chapter-section="2"
                    >
                      2
                    </span>
                     - 
                  </span>
                  <span
                    className="chapter-title"
                  >
                     
                    Mechanics
                     
                  </span>
                  <div
                    className="browse-the-book"
                    onClick={[Function]}
                    role="link"
                  >
                    Browse this Chapter
                  </div>
                </a>
              </div>
            </div>
            <div
              aria-expanded={null}
              aria-hidden={true}
              className="panel-collapse collapse"
              id={undefined}
              role="tabpanel"
            >
              <div
                className="panel-body"
              >
                <div
                  className="section"
                  data-page-id="237"
                  onClick={[Function]}
                >
                  <span
                    className="section-checkbox"
                  >
                    <input
                      checked={false}
                      readOnly={true}
                      type="checkbox"
                    />
                  </span>
                  <span
                    className="chapter-section"
                    data-chapter-section="2.1"
                  >
                    2.1
                  </span>
                  <span
                    className="section-title"
                  >
                     
                    Mechanics in 1 dim
                  </span>
                </div>
                <div
                  className="section"
                  data-page-id="238"
                  onClick={[Function]}
                >
                  <span
                    className="section-checkbox"
                  >
                    <input
                      checked={false}
                      readOnly={true}
                      type="checkbox"
                    />
                  </span>
                  <span
                    className="chapter-section"
                    data-chapter-section="2.2"
                  >
                    2.2
                  </span>
                  <span
                    className="section-title"
                  >
                     
                    Mechanics in 2 dim
                  </span>
                </div>
                <div
                  className="section"
                  data-page-id="239"
                  onClick={[Function]}
                >
                  <span
                    className="section-checkbox"
                  >
                    <input
                      checked={false}
                      readOnly={true}
                      type="checkbox"
                    />
                  </span>
                  <span
                    className="chapter-section"
                    data-chapter-section="2.3"
                  >
                    2.3
                  </span>
                  <span
                    className="section-title"
                  >
                     
                    Mechanics in 3 dim
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="panel-group"
          role="tablist"
        >
          <div
            className="panel panel-default"
            id={null}
          >
            <div
              className="panel-heading"
            >
              <div
                className="chapter-heading panel-title"
                data-chapter-section={3}
              >
                <a
                  aria-controls={undefined}
                  aria-expanded={false}
                  aria-selected={false}
                  className="collapsed"
                  href={undefined}
                  onClick={[Function]}
                  role="tab"
                >
                  <span
                    className="chapter-checkbox"
                  >
                    <span
                      className="tri-state-checkbox unchecked"
                      onClick={[Function]}
                      tabIndex={1}
                    >
                      <i
                        className="tutor-icon fa fa-square-o clickable"
                        onClick={[Function]}
                        role="button"
                        style={
                          Object {
                            "cursor": "pointer",
                          }
                        }
                        type="square-o"
                      />
                    </span>
                  </span>
                  <span
                    className="chapter-number"
                  >
                    Chapter 
                    <span
                      className="chapter-section"
                      data-chapter-section="3"
                    >
                      3
                    </span>
                     - 
                  </span>
                  <span
                    className="chapter-title"
                  >
                     
                    Thermodynamics
                     
                  </span>
                  <div
                    className="browse-the-book"
                    onClick={[Function]}
                    role="link"
                  >
                    Browse this Chapter
                  </div>
                </a>
              </div>
            </div>
            <div
              aria-expanded={null}
              aria-hidden={true}
              className="panel-collapse collapse"
              id={undefined}
              role="tabpanel"
            >
              <div
                className="panel-body"
              >
                <div
                  className="section"
                  data-page-id="240"
                  onClick={[Function]}
                >
                  <span
                    className="section-checkbox"
                  >
                    <input
                      checked={false}
                      readOnly={true}
                      type="checkbox"
                    />
                  </span>
                  <span
                    className="chapter-section"
                    data-chapter-section="3.1"
                  >
                    3.1
                  </span>
                  <span
                    className="section-title"
                  >
                     
                    Thermodynamics in 1 dim
                  </span>
                </div>
                <div
                  className="section"
                  data-page-id="241"
                  onClick={[Function]}
                >
                  <span
                    className="section-checkbox"
                  >
                    <input
                      checked={false}
                      readOnly={true}
                      type="checkbox"
                    />
                  </span>
                  <span
                    className="chapter-section"
                    data-chapter-section="3.2"
                  >
                    3.2
                  </span>
                  <span
                    className="section-title"
                  >
                     
                    Thermodynamics in 2 dim
                  </span>
                </div>
                <div
                  className="section"
                  data-page-id="242"
                  onClick={[Function]}
                >
                  <span
                    className="section-checkbox"
                  >
                    <input
                      checked={false}
                      readOnly={true}
                      type="checkbox"
                    />
                  </span>
                  <span
                    className="chapter-section"
                    data-chapter-section="3.2"
                  >
                    3.2
                  </span>
                  <span
                    className="section-title"
                  >
                     
                    Thermodynamics in 3 dim
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="section-controls panel-footer"
    >
      <div
        className="wrapper"
      >
        <button
          className="btn btn-primary"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Show Questions
        </button>
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
